ConcordiumWallet/Service/MobileWallet.swift

```diff
@@ -102,46 +174,15 @@ class MobileWallet: MobileWalletProtocol {
             
             let data = try IDRequestAndPrivateData(idRequestString)
             return requestPasswordDelegate.requestUserPassword(keychain: keychain)
-                .flatMap { (pwHash) -> AnyPublisher<String, Error> in
-                     var account = AccountDataTypeFactory.create()
-                    do {
-                        // save account name
-                        let initialAccount = self.storageManager.getAccounts().filter {$0.address == ""}.first
-                        var initialAccountName = initialAccount?.name ?? ""
-                        self.storageManager.removeAccount(account: initialAccount)
-                        
-                        // cleanup if there have been other unsucessful tries and if the name was not set
-                        let failedAccounts = self.storageManager.getAccounts(for: identity)
-                        for account in failedAccounts {
-                            if initialAccountName == "" {
-                                initialAccountName = account.name ?? ""
-                            }
-                            self.storageManager.removeAccount(account: account)
-                        }
-                        
-                        // save account
-                        account.name = initialAccountName
-                        account.address = data.initialAccountData.accountAddress
-                        account.encryptedBalanceStatus = .decrypted
-                        account.encryptedAccountData = try self.storageManager.storePrivateAccountKeys(data.initialAccountData.accountKeys,
-                                                                                                       pwHash: pwHash).get()
-                        account.encryptedPrivateKey = try self.storageManager.storePrivateEncryptionKey(data.initialAccountData.encryptionSecretKey,
-                                                                                                        pwHash: pwHash).get()
-                        
-                        _ = try self.storageManager.storeAccount(account)
-                    } catch {
-                        return .fail(error)
-                    }
-                    return self.storageManager.storePrivateIdObjectData(data.privateIDObjectData.value, pwHash: pwHash).onSuccess {
-                        _ = identity.withUpdated(encryptedPrivateIdObjectData: $0)
-                        let newAccount = account.withUpdatedIdentity(identity: identity)
-                        let shieldedAmount = ShieldedAmountTypeFactory.create().withInitialValue(for: newAccount)
-                        _ = try? self.storageManager.storeShieldedAmount(amount: shieldedAmount)
-                    }.publisher.eraseToAnyPublisher()
-            }
-            .map { _ in
-                data.idObjectRequest
-            }.eraseToAnyPublisher()
+                .tryMap { (pwHash) in
```

Comment:
Since you are in a closure you and you reference `self` down below you need to capture it weakly: 

```suggestion
                .tryMap { [weak self] (pwHash) in
```

Reply:
I'm not convinced that this is necessary. (`self` was used previously without being captured weakly.) My understanding is that this is to avoid a retain cycle, leading to a memory leak. I think this cannot happen or doesn't matter for a couple of reasons:
1. `MobileWallet` does not retain the publisher itself, and the caller calls `sink` on the result, so eventually the closure should no longer be retained.
2. The `MobileWallet` instance is retained for the entire lifetime of the application. This means that a retain cycle involving it is irrelevant, since anything that is retained indefinitely by `MobileWallet` will be retained indefinitely whether or not it also retains `MobileWallet`.

On the other hand, `weak self` would mean that we would have to account for the (impossible) case where `self` becomes `nil`.

Reply:
Ah, yeah. I forgot that `MobileWallet` is injected by the `ServicesProvider`. My bad, sorry 🙋🏻‍♂️